controllers/component_customizer.go

```diff
@@ -0,0 +1,291 @@
+	operatorv1 "sigs.k8s.io/cluster-api-operator/api/v1alpha1"
+	"sigs.k8s.io/cluster-api-operator/controllers/genericprovider"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+const (
+	deploymentKind       = "Deployment"
+	namespaceKind        = "Namespace"
+	managerContainerName = "manager"
+	defaultVerbosity     = 1
+)
+
+var (
+	bool2Str = map[bool]string{true: "true", false: "false"}
+)
+
+// customizeObjectsFn apply provider specific customization to a list of manifests.
+func customizeObjectsFn(provider genericprovider.GenericProvider) func(objs []unstructured.Unstructured) ([]unstructured.Unstructured, error) {
+	return func(objs []unstructured.Unstructured) ([]unstructured.Unstructured, error) {
+		results := []unstructured.Unstructured{}
+		for i := range objs {
+			o := objs[i]
+
+			if o.GetKind() == namespaceKind {
+				// filter out namespaces as the targetNamespace already exists as the provider object is in it.
+				continue
+			}
+
+			if o.GetNamespace() != "" {
+				// only set the ownership on namespaced objects.
+				o.SetOwnerReferences(util.EnsureOwnerRef(provider.GetOwnerReferences(),
+					metav1.OwnerReference{
+						APIVersion: operatorv1.GroupVersion.String(),
+						Kind:       provider.GetObjectKind().GroupVersionKind().Kind,
+						Name:       provider.GetName(),
+						UID:        provider.GetUID(),
+					}))
+			}
+
+			if o.GetKind() == deploymentKind {
+				d := &appsv1.Deployment{}
+				if err := scheme.Scheme.Convert(&o, d, nil); err != nil {
+					return nil, err
+				}
+				if provider.GetSpec().Deployment != nil {
+					customizeDeployment(provider.GetSpec(), d)
+				}
+				if err := scheme.Scheme.Convert(d, &o, nil); err != nil {
+					return nil, err
+				}
+			}
+			results = append(results, o)
+		}
+		return results, nil
+	}
+}
+
+// customizeDeployment customize provider deployment base on provider spec input.
+func customizeDeployment(pSpec operatorv1.ProviderSpec, d *appsv1.Deployment) {
+	if pSpec.Deployment != nil {
+		dSpec := pSpec.Deployment
+		if dSpec.Replicas != nil {
+			d.Spec.Replicas = pointer.Int32Ptr(int32(*dSpec.Replicas))
+		}
+		if dSpec.Affinity != nil {
+			d.Spec.Template.Spec.Affinity = dSpec.Affinity
+		}
+		if dSpec.NodeSelector != nil {
+			d.Spec.Template.Spec.NodeSelector = dSpec.NodeSelector
+		}
+		if dSpec.Tolerations != nil {
+			d.Spec.Template.Spec.Tolerations = dSpec.Tolerations
+		}
+
+		for _, pc := range dSpec.Containers {
+			customizeContainer(pc, d)
+		}
+	}
+	// run the customizeManager last so it overrides anything in the deploymentSpec.
+	if pSpec.Manager != nil {
+		for ic, c := range d.Spec.Template.Spec.Containers {
+			if c.Name == managerContainerName {
+				d.Spec.Template.Spec.Containers[ic] = *customizeManager(pSpec.Manager, &d.Spec.Template.Spec.Containers[ic])
+			}
+		}
+	}
+}
+
+// customizeManager customize manager container base on provider spec input.
+func customizeManager(mSpec *operatorv1.ManagerSpec, c *corev1.Container) *corev1.Container {
```

Comment:
We don't actually need a return value from this function do we? We are passing in a pointer and modifying it, returning it and replacing the pointer does nothing